### PR TITLE
5413 - [Notebook] Various visual issues with renaming sections/pages

### LINKED ISSUE #5413: [Notebook] Various visual issues with Notebook navigation
#### Summary

A few visual issues with naming sections or pages in any Notebook (including other notebook types).

#### Expected vs Current Behavior

1. 
**Current**: Section / Page names are truncated while renaming. This is more apparent with a reduced screen width.
**Expected**: Section / Page names should not be truncated while input is active.

https://user-images.githubusercontent.com/9259993/176560917-19162b48-6c00-4caf-ae0d-e4516df4d30e.mov

2. 
**Current**: Hitting 'enter' while renaming a section or page adds a newline instead of confirming entry. Note that the newline is added *before* the text and not after.
**Expected**: Hitting 'enter' should confirm the rename and *not* add a newline. **Question for @charlesh88**: Should newlines be able to be added to section/page names with another key combo such as Shift+Enter?

https://user-images.githubusercontent.com/9259993/176561153-32a1c86e-19ed-4dcd-873d-6bbb3d0db76b.mov

3. 
**Current**: If only one section or page exists in the Notebook, renaming that section or page results in a "blank" that cannot be renamed until context is switched or the section or page is deleted.
**Expected**: After removing the last remaining section or page, the Notebook must always add in a new section or page titled with the correct default value ('Unnamed Section/Page').

https://user-images.githubusercontent.com/9259993/176561364-d837886d-4b1d-4981-837b-5990758a28d8.mov

4. 
Clicking on a section or page to navigate to it immediately puts the name into edit mode, but doesn't assign the correct CSS class for the contenteditable span element. This makes using the nav pane to view pages a very bad experience, and adds a bunch of risk that a section or page may be accidentally renamed.

image

Instead, the user should only be able to edit a previously _selected_ section or page. So to rename a non-selected section or page, they must select it, then click it again. 

image

This mimics how file renaming works in many computer operating systems. This is how it used to work and how it should work now.

#### Steps to Reproduce

See videos above.

#### Environment

Version: 2.0.5
Build Date: Wed Jun 29 2022 12:58:29 GMT-0700 (Pacific Daylight Time)
Revision: e8df2bd437bc2eef193b7eff9c8120a455c0048e
Branch: release/2.0.5

#### Impact Check List

- [ ] Data loss or misrepresented data?
- [x] Regression? Did this used to work or has it always been broken?
- [ ] Is there a workaround available?
- [x] Does this impact a critical component?
- [x] Is this just a visual bug with no functional impact?
- [ ] Does this block the execution of e2e tests?
- [ ] Does this have an impact on Performance?

#### Additional Information
Problems may have been introduced with changes made in 9fbb695379cc695d2a76c23ea3f2566bcc54aa7c.

### PR DESCRIPTION
**Pull request by mariuszr**
Closes #5413 

### Describe your changes:
1. Section / Page names are not truncated while input is active.
2. Hitting `Enter` does confirm the rename and does not add a newline (no multiple line option)
3. After completely removing a section or page title and entering or blurring action returns the default value `'Unnamed Section/Page'`
4. Separate selectability and editability actions for `Section` and `Page` components
5. Additionally, added the `Escape` option that returns the initial value after hitting `Esc`

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?
* [ ] Is this change backwards compatible? For example, developers won't need to change how they are calling the API or how they've extended core plugins such as Tables or Plots.

### Author Checklist

* [x] Changes address original issue?
* [ ] Unit tests included and/or updated with changes?
* [ ] Command line build passes?
* [x] Has this been smoke tested?
* [x] Testing instructions included in associated issue OR is this a dependency/testcase change?

### Reviewer Checklist

* [ ] Changes appear to address issue?
* [ ] Changes appear not to be breaking changes?
* [ ] Appropriate unit tests included?
* [ ] Code style and in-line documentation are appropriate?
* [ ] Commit messages meet standards?
* [ ] Has associated issue been labelled unverified? (only applicable if this PR closes the issue)
* [ ] Has associated issue been labelled bug? (only applicable if this PR is for a bug fix)

### CODE REVIEW

**Comment by mariuszr** [comment]
Hey @jvigliotta and @unlikelyzero, Please review

**Comment by unlikelyzero** [comment]
@mariuszr great work! We're headsdown on a major release and will have this totally reviewed by the end of the week! In the meantime, would you mind adding coverage for some of this functionality so that it doesn't regress in the future?

For the keystroke / key input commands, I think they'd best be suited as e2e tests.

File: src/plugins/notebook/components/SectionComponent.vue [hunk 1]
```diff
@@ -116,16 +116,32 @@ export default {
...
-        updateName(event) {
-            const target = event.target;
-            target.classList.remove('c-input-inline');
-            const name = target.textContent.trim();
+        renameSection(target) {
+            if (!target) {
+                return;
+            }
+
+            const name = target.textContent = target.textContent ? target.textContent.trim().toString() : `Unnamed ${this.sectionTitle}`;
```
Comment (ozyx) [inline src/plugins/notebook/components/SectionComponent.vue:124]:
```suggestion
            const name  = target.textContent ? target.textContent.trim() : `Unnamed ${this.sectionTitle}`;
```

File: src/plugins/notebook/components/SectionComponent.vue [hunk 2]
```diff
@@ -116,16 +116,32 @@ export default {
...
             if (name === '' || this.section.name === name) {
                 return;
             }
 
             this.$emit('renameSection', Object.assign(this.section, { name }));
+        },
+        updateName(event) {
+            const { target, keyCode, type } = event;
+
+            if (keyCode === 27) {
```
Comment (ozyx) [inline src/plugins/notebook/components/SectionComponent.vue:135]: Let's replace these magic numbers with named constants

File: src/plugins/notebook/components/PageComponent.vue [hunk 3]
```diff
@@ -127,16 +128,32 @@ export default {
...
-        updateName(event) {
-            const target = event.target;
-            const name = target.textContent.toString();
-            target.classList.remove('c-input-inline');
+        renamePage(target) {
+            if (!target) {
+                return;
+            }
+
+            const name = target.textContent = target.textContent ? target.textContent.trim().toString() : `Unnamed ${this.pageTitle}`;
```
Comment (ozyx) [inline src/plugins/notebook/components/PageComponent.vue:136]:
```suggestion
            const name = target.textContent ? target.textContent.trim() : `Unnamed ${this.pageTitle}`;
```

**Review by ozyx — CHANGES_REQUESTED** [review]
Looking good! Requesting a couple of small changes.

File: src/plugins/notebook/components/SectionComponent.vue (hunk 1 above)
Comment (mariuszr) [reply src/plugins/notebook/components/SectionComponent.vue:124]: @ozyx, I'll remove `.toString()` function however please note that removing the `target.textContent` assignment with the default value in the case of removing the entire text will cause the header will contain the default value and the component will include the empty string. Is it the expectation?

image

Comment (ozyx) [reply src/plugins/notebook/components/SectionComponent.vue:124]: Ah, good catch. I initially thought this was a typo. In that case, can we separate out the `target.textContent` assignment to make it more clear what's happening here?

Comment (mariuszr) [reply src/plugins/notebook/components/SectionComponent.vue]: I have done some changes in this part of the codes for the Selection and Page components

**Review by ozyx — COMMENTED** [review]
Hi @mariuszr ! Sorry for the delay. I've left a couple suggestions, but overall this is looking really good. The next step would be to cover these changes with some e2e tests. We've recently refactored our e2e framework and folder structure, and added a bunch of documentation. Please check it out and let us know if you have any questions or run into any issues while working on the e2e tests. We are happy to provide any guidance if needed as well. Thanks!

**Comment by mariuszr** [comment]
Hi @ozyx, I'll take a look at the e2e tests.

**Comment by charlesh88** [comment]
@mariuszr This is fantastic, thanks so much! There's an additional regression issue that we found that in master that isn't yet addressed, and I'm hoping you might be able to add a fix to this PR for it. I've added the note to the original issue, but to recap here: 

> Clicking on a section or page to navigate to it immediately puts the name into edit mode, but doesn't assign the correct CSS class for the contenteditable span element:
> 
> image
> 
> Instead, the user should only be able to edit a previously _selected_ section or page. So to rename a non-selected section or page, they must select it, then click it again. 
> 
> image
> 
> This mimics how file renaming works in many computer operating systems. This is how it used to work and how it should work now.

If you can't address this particular issue, let us know and we'll merge this PR in and file a separate issue. But if you can, that'd be even more appreciated. In either case, thanks so much for the contributions, effort and interest.

**Comment by mariuszr** [comment]
Hi @charlesh88, Please note that the plugin header name also behaves with one click option to get the header editable.

image


Providing the double click option to make the component value editable may cause inconsistency in that type of interaction.

**Comment by charlesh88** [comment]
> Hi @charlesh88, Please note that the plugin header name also behaves with one click option to get the header editable.

That is true, but there's an important difference: the main view header name is not really "selectable". The Notebook navigation is more akin to a list of files in a folder view; "selecting" a page or section primarily navigates to that location, in the same way that items in the tree navigate to their views when clicked. The problem we're solving for in the Notebook nav is to _also_ allow a secondary rename action.

**Comment by mariuszr** [comment]
> That is true, but there's an important difference: the main view header name is not really "selectable". The Notebook navigation is more akin to a list of files in a folder view; "selecting" a page or section primarily navigates to that location, in the same way that items in the tree navigate to their views when clicked. The problem we're solving for in the Notebook nav is to _also_ allow a secondary rename action.

@charlesh88 I got your point. Thank you for the explanation. 

I have separated the selectability and editability actions for the `Section` and `Page` components.

**Comment by mariuszr** [comment]
Hi @unlikelyzero and @ozyx Can we create a new issue for creating the E2E test for the task? I will be able to provide it at a later time.

**Comment by ozyx** [comment]
Hi @mariuszr,

Sure! Not a problem, but would you mind adding a couple test stubs so that we don't forget to add coverage here? You can just follow the same pattern [established here](https://github.com/nasa/openmct/blob/master/e2e/tests/functional/plugins/notebook/notebook.e2e.spec.js). No need to implement the tests, just a couple `test.fixme` stubs with some comments describing the test procedures.

Other than that, this PR is just pending a final review from @charlesh88 before we can merge.

Thanks again for all your work on this!

**Review by charlesh88 — APPROVED** [review]
This looks really good, thanks for the changes. I pushed up a very minor tweak that fixes the layout of the list items in the Notebook nav pane to align the affordance arrows:

Screen Shot 2022-08-16 at 10 28 26 AM

With that, this LGTM. Thanks again!

**Review by ozyx — APPROVED** [review]

**Pull request merged**